shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -0,0 +1,275 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.renderer;
+
+import android.annotation.TargetApi;
+import android.graphics.Bitmap;
+import android.graphics.SurfaceTexture;
+import android.os.Build;
+import android.os.Handler;
+import android.support.annotation.NonNull;
+import android.view.Surface;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicLong;
+
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.view.TextureRegistry;
+
+/**
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ *
+ * {@code FlutterRenderer} works in tandem with a provided {@link RenderSurface} to create an
+ * interactive Flutter UI.
+ *
+ * {@code FlutterRenderer} manages textures for rendering, and forwards messages to native Flutter
+ * code via JNI. The corresponding {@link RenderSurface} is used as a delegate to carry out
+ * certain actions on behalf of this {@code FlutterRenderer} within an Android view hierarchy.
+ *
+ * {@link FlutterView} is an implementation of a {@link RenderSurface}.
+ */
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+public class FlutterRenderer implements TextureRegistry {
+
+  private final FlutterJNI flutterJNI;
+  private final long nativeObjectReference;
+  private final AtomicLong nextTextureId = new AtomicLong(0L);
+  private RenderSurface renderSurface;
+
+  public FlutterRenderer(@NonNull FlutterJNI flutterJNI, long nativeObjectReference) {
+    this.flutterJNI = flutterJNI;
+    this.nativeObjectReference = nativeObjectReference;
+  }
+
+  public void attachToRenderSurface(@NonNull RenderSurface renderSurface) {
```

Comment:
What thread assumptions are we making here? lets document them (and/or assert them)

Reply:
That's a good question.  I don't have a great answer off the top of my head.  Historically I think interaction with the JNI occurred from places touching the Android UI which means exclusively the MAIN thread.  But moving forward we may not want to assume that, especially with background execution.

@jason-simmons @bkonyi do either of you have any thoughts or insights with regard to threading in general in the embedding?

Reply:
The embedder Java code should only be called by applications on the main thread.  Callbacks from the engine back into Java should also be invoked only on the main thread.
(I'm not familiar with the details of background execution though.)

If there are any situations where Java code is called on a different thread, then they should be documented specifically.

Reply:
@jason-simmons I was going through and applying the @UiThread annotation to all public methods to codify this fact. Then I saw `getBitmap()` and it got me thinking, are any of these JNI methods long-running?  We won't want to execute any long running operations on the MAIN thread...

Reply:
AFAIK we don't have any APIs that block on I/O.  `getBitmap` is somewhat expensive but should complete its rendering well within the ANR threshold, and it isn't typically used in a critical path.  We could add async version of APIs like this if needed.

Reply:
ping

Reply:
Where? I can't find them in the file... should I be looking elsewhere?
https://github.com/flutter/engine/blob/65420c264582ecdeb318017c3bcc80d10876dd71/shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java#L45